cmake: emu: nsim: Allow all nsim platforms use mdb

wangnuannuan opened the pull request:
Now the twister only supports `nsim_hs_smp` running application with nsim-mdb runner, this commit allows user run target with mdb-nsim runner. ~~What user need to do is to change the simulation type, for example https://github.com/wangnuannuan/zephyr/blob/mainfest-error/boards/arc/nsim/nsim_sem.yaml#L5~~

With this change cmake will choose runner according to `BOARD_DEBUG_RUNNER` or `BOARD_FLASH_RUNNER`. so if add `-DBOARD_FLASH_RUNNER=mdb-nsim` or `-DBOARD_DEBUG_RUNNER=mdb-nsim` to cmake command, the apllication will be run with mdb.

Signed-off-by: Jingru Wang <jingru@synopsys.com>

abrodkin commented:
@wangnuannuan could you please add an example of how that might be used now with your change?

wangnuannuan commented:
When the platform is nsim* in addition to nsim_hs_smp, the default runner is [arc-nsim](https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/arc/nsim/board.cmake#L4), in other words the platform run application with `nsimdrv`. [Here](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/pylib/twister/twisterlib.py#L520) the twister doesn't specify the runner with `--runner`, so twister always runs application with `nsimdrv`. 

My change specific the runner as `mdb-nsim` when the platform's simulation is 'mdb-nsim'.

So if you want `twister` to run applcation with `mdb`, you only need to change the platform's simulation as `mdb-nsim`(in platform's yaml file), then the `twister` will choose `mdb-nsim` as runner

abrodkin commented on cmake/emu/nsim.cmake:
```diff
@@ -1,63 +1,95 @@
+        list(APPEND MDB_OPTIONS ${MDB} -pset=${PSET_NUM} -psetname=core${val})
+        if (val GREATER 0)
+          list(APPEND MDB_OPTIONS -prop=download=2)
+          set(MULTIFILES ${MULTIFILES},core${val})
```
See https://github.com/zephyrproject-rtos/zephyr/pull/36535 on how we order cores (`core0` to be added in the end - rationale is in @IRISZZW's PR).

abrodkin commented on cmake/emu/nsim.cmake:
```diff
@@ -1,63 +1,95 @@
-  MDB
-  mdb
+if("${BOARD_DEBUG_RUNNER}" STREQUAL "mdb-nsim" OR "${BOARD_FLASH_RUNNER}" STREQUAL "mdb-nsim")
+  # mdb is required to run nsim multicore targets
```
Move space changes to a separate commit, see more on that below.

abrodkin reviewed (changes requested):
The idea itself is good, though see my comments on what might be improved.

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,26 +1,62 @@
+endif()
+
+# remove previous .sc.project folder which has temporary settings for MDB.
+set(MDB_OPTIONS rm -rf ${APPLICATION_BINARY_DIR}/.sc.project)
```
I'm curious if it is really working on Windows platform?

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,26 +1,62 @@
 
 if(${CONFIG_SOC_NSIM_HS_SMP})
  set(MDB_ARGS mdb_hs_smp.args)
+elseif(${CONFIG_SOC_NSIM_EM})
+ set(MDB_ARGS mdb_em.args)
+elseif(${CONFIG_SOC_NSIM_EM7D_V22})
+ set(MDB_ARGS mdb_em7d_v22.args)
+elseif(${CONFIG_SOC_NSIM_SEM})
+ set(MDB_ARGS mdb_sem.args)
+elseif(${CONFIG_SOC_NSIM_HS})
+ set(MDB_ARGS mdb_hs.args)
+elseif(${CONFIG_SOC_NSIM_HS6X})
+ set(MDB_ARGS mdb_hs6x.args)
+endif()
```
Should we move it to `boards/arc/nsim/board.cmake`?

We do similar thing for runners arguments in it:
```
if(${CONFIG_SOC_NSIM_EM})
board_runner_args(arc-nsim "--props=nsim_em.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")
elseif(${CONFIG_SOC_NSIM_EM7D_V22})
board_runner_args(arc-nsim "--props=nsim_em7d_v22.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_em7d_v22.args")
elseif(${CONFIG_SOC_NSIM_SEM})
board_runner_args(arc-nsim "--props=nsim_sem.props")
board_runner_args(mdb-nsim "--nsim_args=mdb_sem.args")
elseif(${CONFIG_SOC_NSIM_HS})
board_runner_args(arc-nsim "--props=nsim_hs.props")
 [**snip**]
```

ruuddw commented on boards/arc/nsim/board.cmake:
```diff
@@ -8,21 +8,32 @@ endif()
 board_runner_args(arc-nsim "--props=nsim_em.props")
+set(NSIM_PROPS nsim_em.props)
 board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")
+set(MDB_ARGS mdb_hs_smp.args)
```
With the NSIM_PROPS and MDB_ARGS variables, I suggest to factor out the 
    board_runner_args(arc-nsim "--props=nsim_em.props")
    board_runner_args(mdb-nsim "--nsim_args=mdb_em.args")

and replace them with a single copy using the variables. Something like
    board_runner_args(arc-nsim "--props=${NSIM_PROPS }")
    board_runner_args(mdb-nsim "--nsim_args="${MBD_ARGS}")

ruuddw replied:
I like the change for this, makes it even more compact and scalable, instead of all the hardcoded board names.

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,44 +1,54 @@
-  @${BOARD_DIR}/support/${MDB_ARGS} ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME} &&
-  NSIM_MULTICORE=1 ${MDB} -multifiles=core0,core1 -cmpd=soc -run -cl
+  ${MDB_OPTIONS}
+  ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
   DEPENDS ${logical_target_for_zephyr_elf}
   WORKING_DIRECTORY ${APPLICATION_BINARY_DIR}
   USES_TERMINAL
-  )
+)
```
It would be nice to add printing of MDB command here (i.e. if verbose mode is enabled).
As we construct the command in a quite complex way such logging will allow user to easily get this command to inspect it or run it manually.

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,44 +1,54 @@
-  @${BOARD_DIR}/support/${MDB_ARGS} ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME} &&
-  NSIM_MULTICORE=1 ${MDB} -multifiles=core0,core1 -cmpd=soc -run -cl
+  ${MDB_OPTIONS}
+  ${APPLICATION_BINARY_DIR}/zephyr/${KERNEL_ELF_NAME}
   DEPENDS ${logical_target_for_zephyr_elf}
   WORKING_DIRECTORY ${APPLICATION_BINARY_DIR}
   USES_TERMINAL
-  )
+)
```
It would be nice to add printing of MDB command here (i.e. if verbose mode is enabled).
As we construct the command in a quite complex way such logging will allow user to easily get this command to inspect it or run it manually.

wangnuannuan replied:
I add command to `COMMENT` which is the option of `add_custom_target`. When you run the command `make run` or `ninja run`, it will display the command.

wangnuannuan committed:
cmake: emu: nsim: Allow all nsim platform use mdb

Currently the default runner of nsim_hs_smp is mdb,
the runner of other nsim platform is nsimdrv, user
can't change the default runner unless he uses
west command(--runner).

With this change cmake will choose runner according
to BOARD_DEBUG_RUNNER or BOARD_FLASH_RUNNER. so if
add -DBOARD_FLASH_RUNNER=mdb-nsim or
-DBOARD_DEBUG_RUNNER=mdb-nsim to cmake command,
the apllication will be run with mdb.

Signed-off-by: Jingru Wang <jingru@synopsys.com>

evgeniy-paltsev commented:
Hi Jingru,

I've tried to test your changes, however it looks like we have some issues with them.
When I'm building for nsim_hs_smp the `west flash` command is working, but ninja run isn't.

How to reproduce:
```
west build -b nsim_hs_smp samples/hello_world
west -v flash
 [**snip**]
 [zephyr print] *** Booting Zephyr OS build v2.6.0-rc3-1967-g70458aeb5bf8  ***
 [zephyr print] Hello World! nsim

cd build/
ninja -v run
 [**snip**]
 There is no printing from zephyr!
```

Note that ninja is used by default (instead of make).
Zephyr SDK is 0.13.0-alfa

evgeniy-paltsev commented on boards/arc/nsim/board.cmake:
```diff
@@ -4,27 +4,16 @@ set(EMU_PLATFORM nsim)
-board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=mdb_hs6x_smp.args")
+string(REPLACE "nsim" "mdb" MDB_ARGS "${BOARD}.args")
+if((CONFIG_SOC_NSIM_HS_SMP OR CONFIG_SOC_NSIM_HS6X_SMP))
+board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=${MDB_ARGS}")
```
Just a nitpick: I guess we should add proper formatting here (indentation)

evgeniy-paltsev commented on boards/arc/nsim/board.cmake:
```diff
@@ -4,27 +4,16 @@ set(EMU_PLATFORM nsim)
+if((CONFIG_SOC_NSIM_HS_SMP OR CONFIG_SOC_NSIM_HS6X_SMP))
+board_runner_args(mdb-nsim "--cores=${CONFIG_MP_NUM_CPUS}" "--nsim_args=${MDB_ARGS}")
+else()
+board_runner_args(mdb-nsim "--nsim_args=${MDB_ARGS}")
```
Same here.

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,46 +1,54 @@
+    if(val LESS CONFIG_MP_NUM_CPUS)
+      MATH(EXPR PSET_NUM "${CONFIG_MP_NUM_CPUS}-${val}")
+      MATH(EXPR CORE_NUM "${CONFIG_MP_NUM_CPUS}-${val}-1")
+      if (PSET_NUM GREATER 0)
```
As I remember we should shouldn't add extra space between `if` and `(` in cmake.

evgeniy-paltsev commented on cmake/emu/nsim.cmake:
```diff
@@ -1,46 +1,54 @@
+        list(APPEND MDB_OPTIONS &&)
+      endif()
+      list(APPEND MDB_OPTIONS ${MDB} -pset=${PSET_NUM} -psetname=core${CORE_NUM})
+      if (PSET_NUM GREATER 1)
```
same here.

evgeniy-paltsev reviewed (commented):
Hi Jingru,
several nitpicks from my side :)

evgeniy-paltsev commented on boards/arc/nsim/board.cmake:
```diff
@@ -2,29 +2,18 @@
 if(NOT (CONFIG_SOC_NSIM_HS_SMP OR CONFIG_SOC_NSIM_HS6X_SMP))
-board_set_flasher_ifnset(arc-nsim)
+  board_set_flasher_ifnset(arc-nsim)
 board_set_debugger_ifnset(arc-nsim)
+
+board_runner_args(arc-nsim "--props=${BOARD}.props")
+set(NSIM_PROPS "${BOARD}.props")
```
Just a nitpick: please fix formatting here (indentation) for the whole if block.

evgeniy-paltsev reviewed (commented):
Hi Jingru,

I've re-tested it and now it works fine, thanks. Please find one minor comment, rest of the fings looks good.
